### PR TITLE
Refactor CPU Displacement Filters

### PR DESCRIPTION
**Pull request by klecki**
Shorten & simplify PerSampleCPULoop implementation.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

### CODE REVIEW

**Comment by klecki** [comment]
Build 36320 in progress

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 1]
```diff
@@ -110,6 +112,67 @@ class DisplacementFilter<CPUBackend, Displacement,
   USE_OPERATOR_MEMBERS();
 
  private:
+  // TODO(klecki) We could probably interpolate with something other than float,
+  // for example integer type
+  struct LinearCoefs {
+    LinearCoefs(float rx, float ry) : rx(rx), ry(ry), mrx(1.f - rx), mry(1.f - ry) {}\
```
Comment (klecki) [inline dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: Remove \ at the end of this line.

File: dali/pipeline/operators/displacement/displacement_filter.h [hunk 2]
```diff
@@ -42,6 +42,14 @@ struct HasParam <T, decltype((void) (typename T::Param()), 0)> : std::true_type
 template <typename T>
 struct Point {
   T x, y;
+  template <typename U>
+  Point<U> Cast() {
+    return {static_cast<U>(x), static_cast<U>(y)};
+  }
+  template <typename U>
+  Point<T> operator-(Point<U> r) {
```
Comment (szalpal) [inline dali/pipeline/operators/displacement/displacement_filter.h]: This arithmetic is incompliant with C++ regular type policy. For example:
```
int a=5;
float b=3.2;
auto c = a-b; // decltype(c) == float
```
However, here:
```
Point<int> a = {4,2};
Point<float> b = {5,2};
auto c = a-b; // decltype(c) == Point<int>
```
In my opinion it should at least generate some warning or sth

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 3]
```diff
@@ -110,6 +112,67 @@ class DisplacementFilter<CPUBackend, Displacement,
...
+    LinearCoefs() = default;
+    float rx, ry, mrx, mry;
```
Comment (szalpal) [inline dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: To be compliant with GoogleStyle, there should be `rx_, ry_ etc...`

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 4]
```diff
@@ -110,6 +112,67 @@ class DisplacementFilter<CPUBackend, Displacement,
...
+  };
+
+  template <typename T>
+  bool ShouldTransform(Point<T> p) {
```
Comment (szalpal) [inline dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: argument can be a `const &`

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 5]
```diff
@@ -110,6 +112,67 @@ class DisplacementFilter<CPUBackend, Displacement,
...
+    return p.x >= 0 && p.y >= 0;
+  }
+
+  template <typename T>
+  Index PointToInIdx(Point<T> p, Index H, Index W, Index C) {
```
Comment (szalpal) [inline dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: `p` can be `const &`

Comment (szalpal) [inline dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: `p` could be a `const &`. Same thing applies for few function definitions below

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h (hunk 1 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: Done

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: No, as per GoogleStyle: _Data members of structs, both static and non-static, are named like ordinary nonmember variables. They do not have the trailing underscores that data members in classes have._

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h (hunk 4 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: This is rather simple type, that could also be passed by value. Still, there probably won't be any noticeable difference.

File: dali/pipeline/operators/displacement/displacement_filter.h (hunk 2 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter.h]: You're right, removed it.

**Comment by klecki** [comment]
Build 36426

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 6]
```diff
@@ -110,6 +112,66 @@ class DisplacementFilter<CPUBackend, Displacement,
   USE_OPERATOR_MEMBERS();
 
  private:
+  // TODO(klecki) We could probably interpolate with something other than float,
+  // for example integer type
+  struct LinearCoefs {
+    LinearCoefs(float rx, float ry) : rx(rx), ry(ry), mrx(1.f - rx), mry(1.f - ry) {}
+    LinearCoefs() = default;
```
Comment (pribalta) [inline dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: It is unclear to me what is the benefit of having a default constructor as values seem to be know on initialization. In that case, initializing it in an inconsistent state and filling the members one by one seems an error-prone approach. I get the idea of a POD, but here you can use static typing to ensure correctness

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 7]
```diff
@@ -196,30 +240,14 @@ class DisplacementFilter<CPUBackend, Displacement,
...
-                  // 0, 1
-                  inter_values[2] = in[(yp * W + x) * C + c];
-                  // 1, 1
-                  inter_values[3] = in[(yp * W + xp) * C + c];
-                  out[out_idx + c] = static_cast<T>(
-                      inter_values[0] * mrx * mry +
-                      inter_values[1] * rx * mry +
-                      inter_values[2] * mrx * ry +
-                      inter_values[3] * rx * ry);
+                  load_inputs(in, in_idx + c, next_offsets, inter_values);
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: Will fix

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h [hunk 8]
```diff
@@ -110,6 +112,66 @@ class DisplacementFilter<CPUBackend, Displacement,
   USE_OPERATOR_MEMBERS();
 
  private:
+  // TODO(klecki) We could probably interpolate with something other than float,
+  // for example integer type
+  struct LinearCoefs {
+    LinearCoefs(float rx, float ry) : rx(rx), ry(ry), mrx(1.f - rx), mry(1.f - ry) {}
+    LinearCoefs() = default;
+    float rx, ry, mrx, mry;
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: Done, also I deleted the default constructor.

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h (hunk 6 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: Done.

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h (hunk 4 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: I would prefer to use pass-by-value semantics here.

File: dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h (hunk 5 above)
Comment (klecki) [reply dali/pipeline/operators/displacement/displacement_filter_impl_cpu.h]: As above, I would prefer value semantics over const&.

**Comment by klecki** [comment]
Build 36597

**Pull request merged**